source/common/request_source_plugin_impl.cc

```diff
@@ -0,0 +1,106 @@
+RequestGenerator DummyRequestSourcePlugin::get() {
+  RequestGenerator request_generator = []() {
+    Envoy::Http::RequestHeaderMapPtr header = Envoy::Http::RequestHeaderMapImpl::create();
+    auto returned_request_impl = std::make_unique<RequestImpl>(std::move(header));
+    return returned_request_impl;
+  };
+  return request_generator;
+}
+void DummyRequestSourcePlugin::initOnThread() {}
+
+std::string FileBasedRequestSourceConfigFactory::name() const {
+  return "nighthawk.file-based-request-source-plugin";
+}
+
+Envoy::ProtobufTypes::MessagePtr FileBasedRequestSourceConfigFactory::createEmptyConfigProto() {
+  return std::make_unique<nighthawk::request_source::FileBasedPluginRequestSourceConfig>();
+}
+
+RequestSourcePluginPtr FileBasedRequestSourceConfigFactory::createRequestSourcePlugin(
+    const Envoy::Protobuf::Message& message, Envoy::Api::Api& api) {
+  const auto& any = dynamic_cast<const Envoy::ProtobufWkt::Any&>(message);
+  nighthawk::request_source::FileBasedPluginRequestSourceConfig config;
+  Envoy::MessageUtil::unpackTo(any, config);
+  return std::make_unique<FileBasedRequestSourcePlugin>(config, api);
+}
+
+REGISTER_FACTORY(FileBasedRequestSourceConfigFactory, RequestSourcePluginConfigFactory);
+
+FileBasedRequestSourcePlugin::FileBasedRequestSourcePlugin(
+    const nighthawk::request_source::FileBasedPluginRequestSourceConfig& config,
+    Envoy::Api::Api& api)
+    : RequestSourcePlugin{api}, uri_(config.uri()), file_path_(config.file_path()) {
+  Envoy::MessageUtil util;
+  util.loadFromFile(file_path_, options_list_, Envoy::ProtobufMessage::getStrictValidationVisitor(),
+                    api_, true);
+}
+
+RequestGenerator FileBasedRequestSourcePlugin::get() {
+  RequestOptionsIterator iterator = options_list_.options().begin();
+  request_iterators_.push_back(iterator);
+  RequestOptionsIterator* temp = &request_iterators_.back();
+  RequestGenerator request_generator = [this, temp]() mutable {
+    nighthawk::client::RequestOptions request_option = **temp;
+    *temp = std::next(*temp);
+    Envoy::Http::RequestHeaderMapPtr header = Envoy::Http::RequestHeaderMapImpl::create();
+    header->setPath(uri_.path());
+    header->setHost(uri_.hostAndPort());
+    header->setScheme(uri_.scheme() == "https" ? Envoy::Http::Headers::get().SchemeValues.Https
+                                               : Envoy::Http::Headers::get().SchemeValues.Http);
+    header->setMethod(envoy::config::core::v3::RequestMethod_Name(request_option.request_method()));
+    const uint32_t content_length = request_option.request_body_size().value();
+    if (content_length > 0) {
+      header->setContentLength(content_length);
+    }
+    for (const auto& option_header : request_option.request_headers()) {
+      auto lower_case_key = Envoy::Http::LowerCaseString(std::string(option_header.header().key()));
+      header->remove(lower_case_key);
+      header->addCopy(lower_case_key, std::string(option_header.header().value()));
```

Comment:
I wonder, Isn't there a `setCopy()` around instead of the remove/add? Also -- should this just add? Or ... maybe what happens should be part of the options spec (replace/remove/append)?

Reply:
There is a setCopy(). I'm not sure about making it an option, but I'll spin off a separate issue for adding that so I can get this PR in.

Reply:
Sounds good